### ISSUE TITLE
Exceptions thrown for pretty much any Keras imported model in deeplearning4j-1.0.0beta

### ISSUE THREAD

**Issue by octocode686** [deeplearning4j/deeplearning4j#6527]
Hi,

It seems that even the simplest keras models are not translating well into DL4J for some reason. The stack trace varies slightly but is always along the lines of 
Exception in thread "main" java.lang.ClassCastException: java.util.LinkedHashMap cannot be cast to java.util.List
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:100)
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:60)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildSequential(KerasModelBuilder.java:147)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasSequentialModelAndWeights(KerasModelImport.java:181)
	etc

This is taken from the dl4j-examples and the SimpleSequentialImport class. 

Am I doing something fundamentally wrong for this to occur? I have literally failed to load every single model I have tried!
It seems that the KerasSequentialModel is trying to cast the input as a List, but it is coming in as a LinkedHashMap. Could the linked hash map be a default when java doesn't know how to unpack the h5 model? I am using the importKerasSequentialModelAndWeights method for the import of a complete model.
Thanks in advance.

**Comment by AlexDBlack** [comment]
Try with 1.0.0-beta2. There were a number of fixes there. Also note that we have a ton of unit tests for this functionality, and has been pretty robust in our experience: https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j/deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras

If you're still having trouble with any nets on 1.0.0-beta2, we'll need to see some code and networks to help debug this.

**Comment by octocode686** [comment]
Hi Alex,
Thank you for the swift response, it is appreciated. I tried and the problem remains the same, I am thinking that I'm missing something crucial.
I am trying to run Max Pumperla's SimpleSequentialMlpImport class example from dl4j-examples using the model attached to this post. Thank you again for having a look at this, it seems quite odd that I can't seem to make it work! Below is the exception thrown as a result of the code's execution. 

Exception in thread "main" java.lang.ClassCastException: java.util.LinkedHashMap cannot be cast to java.util.List
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:102)
	at org.deeplearning4j.nn.modelimport.keras.KerasSequentialModel.<init>(KerasSequentialModel.java:61)
	at org.deeplearning4j.nn.modelimport.keras.utils.KerasModelBuilder.buildSequential(KerasModelBuilder.java:320)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasSequentialModelAndWeights(KerasModelImport.java:180)
	at package org.deeplearning4j.examples.modelimport.keras.basic.SimpleSequentialMlpImport.main(SimpleSequentialMlpImport.java:70)



/*
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/
package org.deeplearning4j.examples.modelimport.keras.basic;

import org.deeplearning4j.nn.api.Layer;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.layers.DenseLayer;
import org.deeplearning4j.nn.conf.layers.LossLayer;
import org.deeplearning4j.nn.modelimport.keras.KerasModelImport;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.nd4j.linalg.activations.Activation;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.io.ClassPathResource;
import org.nd4j.linalg.lossfunctions.LossFunctions;

 * @author Max Pumperla
 */
public class SimpleSequentialMlpImport {


    public static void main(String[] args) throws Exception {
        final String SIMPLE_MLP = new ClassPathResource("modelimport/keras/simple_mlp.h5").getFile().getPath();

        MultiLayerNetwork model = KerasModelImport.importKerasSequentialModelAndWeights(SIMPLE_MLP, true);

        // Test basic inference on the model.
        INDArray input = Nd4j.create(256, 100);
        INDArray output = model.output(input);
        model.fit(input, output);
        assert model.conf().getOptimizationAlgo().equals(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT);

        Layer first = model.getLayer(0);
        DenseLayer firstConf = (DenseLayer) first.conf().getLayer();
        assert firstConf.getActivationFn().equals(Activation.RELU.getActivationFunction());
        assert firstConf.getNIn() == 100;
        assert firstConf.getNOut() == 64;

        Layer second = model.getLayer(1);
        DenseLayer secondConf = (DenseLayer) second.conf().getLayer();
        assert secondConf.getActivationFn().equals(Activation.SOFTMAX.getActivationFunction());
        assert secondConf.getNIn() == 64;
        assert secondConf.getNOut() == 10;

        Layer loss = model.getLayer(2);
        LossLayer lossConf = (LossLayer) loss.conf().getLayer();
        assert lossConf.getLossFn().equals(LossFunctions.LossFunction.MCXENT);
    }
}

[keras_fasttext_model_full.zip](https://github.com/deeplearning4j/deeplearning4j/files/2449649/keras_fasttext_model_full.zip)

**Comment by octocode686** [comment]
I would like to add that I just managed to run the basic MLP model from the examples written by Max. 
Perhaps the model I've attached to my previous post has not been serialized properly?

**Comment by Kissy** [comment]
I reproduce the exception compiling with Keras 2.2.3 and importing with 1.0.0-beta2.
It seems that the keras field for "config" (retrieve here modelConfig.get(config.getModelFieldConfig())) has changed from list of object in version 2.0.X to object representing map.

Keras 2.0.9 :
`{
  "backend": "tensorflow",
  "keras_version": "2.0.9",
  "class_name": "Sequential",
  "config": [
    {
      "class_name": "Dense",
      "config": {
        "kernel_initializer": {
          "class_name": "VarianceScaling",
          "config": {
            "distribution": "uniform",
            "scale": 1.0,
            "seed": null,
            "mode": "fan_avg"
          }`

Keras 2.2.3 :
`{
  "backend": "tensorflow",
  "keras_version": "2.2.3",
  "class_name": "Sequential",
  "config": {
    "name": "sequential_1",
    "layers": [
      {
        "class_name": "Dense",
        "config": {
          "name": "dense_1",
          "trainable": true,
          "batch_input_shape": [
            null,
            100
          ],`

**Comment by octocode686** [comment]
Hi Guillaume,
You confirmed my suspicions, for which I am grateful. It seems that this is indeed the problem and downgrading Keras to 2.0.9 and using the models produced by that version worked smoothly.
Thanks again for everyone's time!

**Comment by Kissy** [comment]
You're welcome. 
I am not sure if there is an issue for making dl4j compatible with Keras 2.3.X

**Comment by octocode686** [comment]
Indeed it probably is. A new thread would perhaps be adequate?

**Comment by maxpumperla** [comment]
@octocode686 @Kissy I'm going to close this one in favor of #6550. Sorry, I was unaware of the breaking change in model serialization.
